### PR TITLE
[lang] Add matrix support for Dynamic SNode

### PR DESCRIPTION
**Pull request by lin-hitonami**
Issue: #5420

### CODE REVIEW

File: tests/python/test_dynamic.py [hunk 1]
```diff
@@ -283,3 +264,51 @@ def make_list():
             assert x[i, j].b == i * j * 10000 % 65536
             assert x[i, j].c == i * j * 100000000 % 4294967296
             assert x[i, j].d == i * j * 10000000000
+
+
+@test_utils.test(require=ti.extension.sparse, exclude=[ti.metal])
```
Comment (jim19930609) [inline tests/python/test_dynamic.py:269]: Can you try to add another test with `real_matrix=True, real_matrix_scalarize=True`? Something like:
```
@test_utils.test(require=ti.extension.sparse, exclude=[ti.metal], real_matrix=True, real_matrix_scalarize=True)
def test_append_matrix_scalarize():
     _test_append_matrix()
```

I can help with the fix if it fails.

File: tests/python/test_dynamic.py [hunk 2]
```diff
@@ -283,3 +264,80 @@ def make_list():
...
+def test_append_matrix():
+    _test_append_matrix()
+
+
+@test_utils.test(require=ti.extension.sparse,
+                 exclude=[ti.metal],
+                 real_matrix=True,
```
Comment (ailzhang) [inline tests/python/test_dynamic.py:298]: parametrize?

**Review by jim19930609 — APPROVED** [review]

File: tests/python/test_dynamic.py (hunk 2 above)
Comment (lin-hitonami) [reply tests/python/test_dynamic.py:298]: Can pytest.parametrize be used on the decorator?

Comment (ailzhang) [reply tests/python/test_dynamic.py:298]: @lin-hitonami do you mean https://github.com/taichi-dev/taichi/blob/master/tests/python/test_ndarray.py#L38? ;)

Comment (lin-hitonami) [reply tests/python/test_dynamic.py:298]: I mean we can't use it on parameters in `@test_utils.test()`. (Do you mean we can parametrize `real_matrix=True`?)

**Pull request merged**